### ISSUE TITLE
Potential GPU Performance Bottleneck?

### ISSUE THREAD

**Issue by yudongqiu** [pandegroup/openmm#1567]
Dear Developers,

I notices that when running `openmm/examples/benchmark.py` on CUDA platform, the GPU utilization was kept at about 85% during the amoebapme test, which makes me courious if there is a performace bottleneck that is holding the GPU back. I found that the python process running openmm uses less than 40% of a single CPU core, and minimal CPU memory and GPU memory. The PCIe bandwidth usage was 0% during the run, and the disk was left free of use.

Apart from the amoebapme test, all other tests (gbsa, rf, pme, amoebagk) basically utilize 100% of the GPU. Direct polarization also avoids the bottleneck, also showing 100% GPU usage. 

The tests were ran on GTX 980ti and GTX 1080 graphic cards. As a result of the bottleneck, GTX 1080 performance were only 8% higher than GTX 980ti for the amoebapme test, but in other cases when the GPU was utilized near 100%, GTX 1080 showed at least 30% higher performance than GTX 980ti.

This bottleneck also appeared in our real applications with AMOEBA forcefield and mutual polarization. Therefore I think it would be nice if we can find a way to avoid it.

Thanks,

Yudong Qiu

**Comment by peastman** [comment]
Thanks for pointing that out.  It's entirely possible it could be better optimized.  I'm pretty sure the relevant code is the implementation of DIIS for iterating the induced dipoles.  At each iteration, it needs to build and invert a small matrix.  Currently we build the matrix on the GPU, but then download it to the CPU to invert.  At the time I wrote it, there wasn't any convenient library for doing the matrix operations on the GPU, and I figured the performance cost would be fairly small.  But CUDA 7.0 introduced the cuSolver library which could probably be used for this, and that would eliminate the cost of sending the matrix back to the CPU.

**Comment by swails** [comment]
It'll be interesting to see if the SVD we use to solve the error matrix in DIIS will actually run faster on the GPU.  The matrix is such low-dimensional that I suspect the answer is no for highly optimized math libraries like the MKL ([based on an old paper about an SVD implementation on the GPU](http://www.pascal-man.com/navigation/faq-java-browser/GPU/Sheetal09Singular.pdf)).

I wonder if using the MKL would improve performance... (do we even use BLAS routines?)

We also do a lot more PME calculations with mutual induced dipoles, which means more dipole spreading on the grid (which means more atomic ops to global memory).  I can't think of anything else that really differs from a potential non-compute bottleneck for performance between mutual and direct dipoles, though...

**Comment by peastman** [comment]
The time to do the SVD should be negligible.  It's the communication that's the problem.  Once we build the matrix, we have to send it to the CPU, then do the SVD to determine the coefficients, then transfer them back again.  If we assume each transfer has a latency of 10 us and it performs 10 iterations per time step, that's 200 us per time step.  It's not a huge overhead, but not negligible either.

**Comment by yudongqiu** [comment]
Thanks a lot for the explanation! I think it will be exciting to see how much the performance could be improved if the DIIS equations can be solved directly on the GPU.

**Comment by peastman** [comment]
I've been trying to implement this.  The problem I'm running into is that, as far as I can tell, the matrix inversion really has to be done in double precision.  Here's a typical example of the sort of matrix we're trying to invert:

```
[0, -1, -1, -1],
[-1, 3.17042e-07, -7.97217e-08, 3.03149e-09],
[-1, -7.97217e-08, 2.48617e-08, -1.21446e-09],
[-1, 3.03149e-09, -1.21446e-09, 1.55487e-10]
```

In double precision, cuSolver perfectly matches the results of the existing solver.  But in single precision, it gives completely wrong results.  The matrix elements vary too widely in magnitude.

I need to figure out whether it's worth doing anyway.  But on GeForce GPUs, I think it's likely it will end up being slower than what we're doing now.

**Comment by jchodera** [comment]
This matrix looks terribly ill-conditioned. Is there a trick where we can
multiply it by known matrices to improve the condition number first?

**Comment by kyleabeauchamp** [comment]
Does the 4x4 matrix always have "special structure" e.g. patterns of ones and zeros?  If so, I wonder if the analytical inverse (e.g. mathematica sympy) might help avoid doing operations in floating point?  Also: regarding MKL, long ago we found that hand-tuned code outperformed MKL for low-dimensional matrix multiplies for RMSD calculation, might be similar here.

**Comment by swails** [comment]
Have you tried going back to SOR to see if that eliminates the GPU under-utilization?  If it doesn't, the DIIS probably isn't the issue, anyway, right?

**Comment by swails** [comment]
> Does the 4x4 matrix always have "special structure" e.g. patterns of ones and zeros? If so, I wonder if the analytical inverse (e.g. mathematica sympy) might help avoid doing operations in floating point? Also: regarding MKL, long ago we found that hand-tuned code outperformed MKL for low-dimensional matrix multiplies for RMSD calculation, might be similar here.

It's not always 4x4.  I think the current implementation saves up to 15 error vectors, so the size is not the same each iteration (if I'm remembering correctly, that is)

**Comment by peastman** [comment]
That's correct.  The size of the matrix increases by 1 on each iteration up to a maximum that we set.  And yes, the top row and left column are always the same.  The remaining elements are dot products of the error vectors from different iterations, which tend to be pretty small.

**Comment by jchodera** [comment]
There are some special matrix inverse relationships for matrices of special structure---like block matrices---that could potentially be exploited to reduce the inversion to just the better-conditioned submatrix block. Is there I simple explanation of the mathematics defining the matrix that must be inverted you can point us to?

Another possibility is to work on the matrix inverse itself and iteratively update it as the new error vector dot products are added, much like L-BFGS avoids explicitly inverting the Hessian matrix by successively applying an update rule (generally implicitly) to the inverse Hessian.

**Comment by peastman** [comment]
Here's an explanation of where the matrix comes from: http://vergil.chemistry.gatech.edu/notes/diis/node2.html

**Comment by yudongqiu** [comment]
May I please ask if you have considered the "Batched Solver" from cuBLAS?
[https://developer.nvidia.com/rdp/assets/cuda-batched-solver-tgz](url)
I haven't tried it before but I saw that it can support matrices with dimensions between 2~77 in double precision which seems to be suitable for our problem here.
From my understanding of the issue, the GPU solver does not need to out-perform the CPU ones for the small matrices, but it just need to consume less time than that combined for back and forward transfer latencies and the CPU solver. I feel it's pretty interesting that for regular performance benchmarks people consider the transfer latencies to be on the GPU side, but in our case it's the opposite, so I have a good confidence in this new approach.

**Comment by peastman** [comment]
I'll take a look at that.  cuSolver seems to be disastrously inefficient for this case.  Using it slows down the simulation by about a factor of 3.  And when I look at it in the profiler, I see that it's launching tons of kernels, each of which executes for an insignificant amount of time, but just the overhead from them becomes the main bottleneck.  It seems to be written for working with large matrices, not small ones.

Another option would be to just take the SVD code from JAMA, port it to CUDA, and run it as a single thread block.  That might not be too difficult.

**Comment by jlmaccal** [comment]
> Another option would be to just take the SVD code from JAMA, port it to CUDA, and run it as a single thread block. That might not be too difficult.
> 

I faced the same issue a few years back implementing RDC restraints. All of the CUDA libraries seemed optimized for huge matrices. I currently round-trip things to the CPU, so a GPU SVD routine optimized for small matrices would benefit us as well.

**Comment by peastman** [comment]
I've been trying lots of things, but all of them end up being slower than just sending the matrix to the CPU.  I suspect it's possible to write GPU matrix routines that work efficiently on small matrices, but I don't have expertise to do it myself, and I haven't found any existing ones.

I did manage to speed it up a tiny bit by switching from SVD to LU.  I was using SVD because I was worried about the matrix being ill conditioned, but it turned out in practice that wasn't an issue and it was always including all singular values.  But since the computation time was already smaller than the communication time, this only gives a tiny speedup (around 1% in my tests).

**Comment by yudongqiu** [comment]
Thanks a lot for the update. I just ran the benchmark after applying the lastest updates, and the amoebapme spped under single precision jumped up from 10 ns/day to about 15 ns/day with the Titan X Pascal!

**Comment by peastman** [comment]
I finally found a better solution (#1640).  I figured out a way to precondition the matrix so it could be solved in single precision.  Running LU decomposition as a single thread on the GPU is now faster than sending the matrix to the CPU.  There's probably a way to use multiple threads to do it even faster, but that kernel now takes less than 3% of the total GPU time in my test, so it's not a big deal.  I might still look into it at some point though.
